--- labyrinth1.c	2025-11-23 20:19:33.364614229 +0300
+++ labyrinth2.c	2025-11-23 20:19:33.430812709 +0300
@@ -84,34 +84,39 @@
     } while ((double)broken_walls / total_walls > 0.85);
 }
 
-void print_labyrinth(uint8_t *labyrinth, size_t size) {
+void print_labyrinth(uint8_t *labyrinth, size_t size, char passage, char wall) {
     for (int x = 0; x < size; x++)
-        printf("##");
-    printf("#\n");
+        printf("%c%c", wall, wall);
+    printf("%c\n", wall);
     for (int y = 0; y < size; y++) {
-        printf("#.");
+        printf("%c%c", wall, passage);
         for (int x = 1; x < size; x++) {
-            char wall = CELL(labyrinth, size, y, x) & LEFT ? '#' : '.';
-            printf("%c.", wall);
+            char wall_or_passage = CELL(labyrinth, size, y, x) & LEFT ? wall : passage;
+            printf("%c%c", wall_or_passage, passage);
         }
-        printf("#\n");
+        printf("%c\n", wall);
         for (int x = 0; x < size; x++) {
-            char wall = CELL(labyrinth, size, y, x) & DOWN ? '#' : '.';
-            printf("#%c", wall);
+            char wall_or_passage = CELL(labyrinth, size, y, x) & DOWN ? wall : passage;
+            printf("%c%c", wall, wall_or_passage);
         }
-        printf("#\n");
+        printf("%c\n", wall);
     }
 }
 
 int main(int argc, char *argv[]) {
-    if (argc != 2) {
-        fprintf(stderr, "Usage: %s <labyrinth size>\n", argv[0]);
+    if (argc != 3) {
+        fprintf(stderr, "Usage: %s '<passage symbol><wall symbol>' <labyrinth size>\n", argv[0]);
         exit(1);
     }
     char *endptr = NULL;
-    long size = strtol(argv[1], &endptr, 0);
-    if (endptr == argv[1] || *endptr != '\0' || size < 1) {
-        fprintf(stderr, "Wrong labyrinth size specified: \"%s\"\n", argv[1]);
+    int argv1_len = strlen(argv[1]);
+    if (argv1_len != 2) {
+        fprintf(stderr, "The first parameter must consist of 2 characters, got %d\n", argv1_len);
+        exit(1);
+    }
+    long size = strtol(argv[2], &endptr, 0);
+    if (endptr == argv[2] || *endptr != '\0' || size < 1) {
+        fprintf(stderr, "Wrong labyrinth size specified: \"%s\"\n", argv[2]);
         exit(1);
     }
     uint8_t *labyrinth = alloc_labyrinth(size);
@@ -120,7 +125,7 @@
         exit(1);
     }
     gen_labyrinth(labyrinth, size);
-    print_labyrinth(labyrinth, size);
+    print_labyrinth(labyrinth, size, argv[1][0], argv[1][1]);
     free(labyrinth);
     return 0;
 }
