--- labyrinth2.c	2025-11-23 20:19:33.430812709 +0300
+++ labyrinth3.c	2025-11-23 20:19:33.496613637 +0300
@@ -38,8 +38,7 @@
     return -1;
 }
 
-void gen_labyrinth(uint8_t *labyrinth, size_t size) {
-    int seed = 1231242;
+void gen_labyrinth(uint8_t *labyrinth, size_t size, unsigned seed) {
     int total_walls = (size - 1) * size * 2;
     int broken_walls = 0;
     do {
@@ -104,28 +103,40 @@
 }
 
 int main(int argc, char *argv[]) {
-    if (argc != 3) {
-        fprintf(stderr, "Usage: %s '<passage symbol><wall symbol>' <labyrinth size>\n", argv[0]);
+    if (argc != 4) {
+        fprintf(stderr, "Usage: %s <random seed> '<passage symbol><wall symbol>' <labyrinth size>\n", argv[0]);
         exit(1);
     }
+    char *random_seed_argv = argv[1];
+    char *syms_argv = argv[2];
+    char *size_argv = argv[3];
+
     char *endptr = NULL;
-    int argv1_len = strlen(argv[1]);
-    if (argv1_len != 2) {
-        fprintf(stderr, "The first parameter must consist of 2 characters, got %d\n", argv1_len);
+    long seed = strtol(random_seed_argv, &endptr, 0);
+    if (endptr == random_seed_argv || *endptr != '\0') {
+        fprintf(stderr, "Wrong random seed specified: \"%s\"\n", random_seed_argv);
+        exit(1);
+    }
+    endptr = NULL;
+    int syms_len = strlen(syms_argv);
+    if (syms_len != 2) {
+        fprintf(stderr, "The first parameter must consist of 2 characters, got %d\n", syms_len);
         exit(1);
     }
-    long size = strtol(argv[2], &endptr, 0);
-    if (endptr == argv[2] || *endptr != '\0' || size < 1) {
-        fprintf(stderr, "Wrong labyrinth size specified: \"%s\"\n", argv[2]);
+    long size = strtol(size_argv, &endptr, 0);
+    if (endptr == size_argv || *endptr != '\0' || size < 1) {
+        fprintf(stderr, "Wrong labyrinth size specified: \"%s\"\n", size_argv);
         exit(1);
     }
+
     uint8_t *labyrinth = alloc_labyrinth(size);
     if (labyrinth == NULL) {
         fprintf(stderr, "Failed to alloc labyrinth: %s\n", strerror(errno));
         exit(1);
     }
-    gen_labyrinth(labyrinth, size);
-    print_labyrinth(labyrinth, size, argv[1][0], argv[1][1]);
+    gen_labyrinth(labyrinth, size, seed);
+    print_labyrinth(labyrinth, size, syms_argv[0], syms_argv[1]);
+
     free(labyrinth);
     return 0;
 }
